org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/service/ServiceUseExtensionTest.java

```diff
@@ -0,0 +1,359 @@
+/*
+ * Copyright (c) OSGi Alliance (2019). All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.osgi.test.junit5.service;
+
+import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import java.util.Hashtable;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import org.assertj.core.api.SoftAssertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ExtensionContext.Store;
+import org.mockito.stubbing.Answer;
+import org.osgi.framework.InvalidSyntaxException;
+import org.osgi.test.common.tracking.TrackServices;
+import org.osgi.test.junit5.ExecutorExtension;
+import org.osgi.test.junit5.ExecutorParameter;
+import org.osgi.test.junit5.types.Foo;
+import org.osgi.test.junit5.types.MockStore;
+
+@ExtendWith(ExecutorExtension.class)
+public class ServiceUseExtensionTest {
```

Comment:
If `ServiceUseRule` and `ServiceUseExtension` had a common base class, then their respective test classes could too. That way you can avoid duplicating the test code between each and also give a greater guarantee that their behaviour is consistent.

Reply:
that would have to be a new module right? You can't share code from a module that has it in `src/test/main` right? I think that's wonky!?! I'm open to ideas.

Reply:
The other issue is exposing the junit apis to each other. I'm happy keeping the code somewhat duplicative in that there is no junit 3/4 api exposure to junit5 testing and vice-versa.

Reply:
I filed an issue for this.